File: upwind3.jl

type Upwind
    ndims::Int
    velocity::Array
    lengths::Array
    numCells::Array
    deltas::Array
    upDirection::Array
    offsetMat::Array
    numCellsExt::Array
    coeff::Array
    ntot::Int
    dimProd::Array
    f::Array
end

function advect!(up::Upwind,deltaTime)

    # copy
    oldF = deepcopy(up.f)

    # index set for each cell
    inds = zeros(Integer, up.ndims, up.ntot)
    
    for j = 1:up.ndims
        inds[j,:] = collect(1:up.ntot)
        inds[j,:] /= up.dimProd[j]
        inds[j,:] %= up.numCells[j]
    end

    indsUp = deepcopy(inds)
    
    # Update the field in each spatial direction
    for j = 1:up.ndims

        # Apply offset
        indsUp[j,:] += up.upDirection[j]
        indsUp[j,:] %= up.numCells[j]

        #compute flat indices corresponding to the offset index sets
        flatIndsUp = dot(up.dimProd, indsUp)

        #update 
        up.f -= (deltaTime * up.coeff[j])*(oldF[flatIndsUp] - oldF)

        # Reset
        indsUp[j,:] = deepcopy(inds[j,:])
    end
end

function getFlatIndex(up::Upwind,inds)
    ans = Int(dot(up.dimProd,inds))
    return ans
end

function getIndexSet(up::Upwind,flatIndex)
    res = zeros(Int,up.ndims)
    for i = 0:up.ndims-1
        res[i+1] =  mod(Int(floor(flatIndex/up.dimProd[i+1])), up.numCells[i+1])
    end
    return res
end

function  saveVTK(up::Upwind)

end

function checksum(up::Upwind)
    return sum(up.f)
end



#Beginning of main program
ndims = 3
#For quick development, I will hard code now
#Later I will take these as parameters.
#ncells = 1
ncells = parse(Int, ARGS[1])
#numTimeSteps = 1
numTimeSteps = parse(Int, ARGS[2])

#Same resolution in each direction.
numCells = [ncells,ncells,ncells]
floatingnumCells = [convert(AbstractFloat,ncells),convert(AbstractFloat,ncells),convert(AbstractFloat,ncells)]
println(numCells)



#Initalizing vector velocity and lengths
velocity = ones(Float64,ndims)
#println(velocity)
lengths   = ones(Float64,ndims)
#println(lengths)


#computer dt
courant = 0.1
dt = Inf
dx = 0.00e0
for i = 1:ndims
    dx = lengths[i]/float(numCells[i])
    dt = min((courant*dx)/velocity[i],dt)
end
#println(dx)
#println(dt)

#This is where upwind class is constructed in C++ and python version of this code
v = velocity
deltas = zeros(Float64,ndims)
upDirection = zeros(Integer,ndims)
floatupDirection = zeros(Float64,ndims)
offsetMat = eye(ndims,ndims)
numCellsExt = zeros(Float64,ndims,ndims)
BLAS.ger!(1.0,floatingnumCells,ones(Float64,ndims),numCellsExt)

ntot = 1
for i in 1:ndims
    upDirection[i] = -1
    floatupDirection[i] = -1.00e0
    if velocity[i] < 0.00
        upDirection[i] = +1
        floatupDirection[i] = +1.00e0
    end
    deltas[i] = lengths[i]/numCells[i]
    offsetMat[i,i] = upDirection[1]
    ntot *= numCells[i]
end
#println("upDirection= ",upDirection)
#println("ntot= ",ntot)
#println("delta= ",deltas)
dimProd= ones(Integer,ndims)
for i = ndims-2:-1:0
    dimProd[i+1] = dimProd[i+2]*numCells[i+2]
end
#println("dimProd= ",dimProd)

coeff = zeros(Float64,ndims)
for i in 1:ndims
    coeff[i] = velocity[i] * floatupDirection[i] / deltas[i]
end 

f = zeros(ntot)
f[1] = 1.00e0
up = Upwind(ndims,velocity,lengths,numCells,deltas,upDirection,offsetMat,numCellsExt,coeff,ntot,dimProd,f)
#println(up)


#Do the time evolution of the upwind class
for i = 1:numTimeSteps
   advect!(up,dt) 
end 
println("Time evolution complete")

#Do the checksum
#println("Last f= ",up.f)
println("Check_sum= ",checksum(up))

File: upwind3.py
#/usr/bin/env python

import numpy
import copy

class Upwind: 

  def __init__(self, velocity, lengths, numCells):
    self.numCells = numCells
    self.ndims = len(velocity)
    self.deltas = numpy.zeros( (self.ndims,), numpy.float64 )
    self.upDirection = numpy.zeros( (self.ndims,), numpy.float64 )
    self.v = velocity
    self.lengths = lengths
    self.ntot = 1
    self.offsetMat = numpy.identity(self.ndims, numpy.int)
    self.numCellsExt = numpy.outer(self.numCells, numpy.ones((self.ndims,), numpy.int))
    for j in range(self.ndims):
      self.upDirection[j] = -1
      if velocity[j] < 0.: self.upDirection[j] = +1
      self.deltas[j] = lengths[j] / numCells[j]
      self.offsetMat[j, j] = self.upDirection[j]
      self.ntot *= numCells[j]

    self.dimProd = numpy.ones( (self.ndims,), numpy.int )
    for i in range(self.ndims - 2, -1, -1):
        # last index varies fastest
        self.dimProd[i] =  self.dimProd[i + 1] * self.numCells[i + 1]

    self.coeff = self.v * self.upDirection / self.deltas

    # initializing the field
    self.f = numpy.zeros( (self.ntot,), numpy.float64 )
    # initialize lower corner to one
    self.f[0] = 1

  def advect(self, deltaTime):

    # copy 
    oldF = self.f.copy()

    # array of index sets for each cell
    inds = numpy.zeros( (self.ndims, self.ntot), numpy.int )
    for j in range(self.ndims):
      inds[j, :] = numpy.arange(self.ntot)
      inds[j, :] /= self.dimProd[j]
      inds[j, :] %= self.numCells[j]

    indsUp = inds.copy()

    # update the field in each spatial direction
    for j in range(self.ndims):

      # apply offset
      indsUp[j, :] += self.upDirection[j]
      indsUp[j, :] %= self.numCells[j]

      # compute flat indices corresponding to the offset index sets
      flatIndsUp = numpy.dot(self.dimProd, indsUp)

      # update
      self.f -= (deltaTime * self.coeff[j]) * (oldF[flatIndsUp] - oldF)

      # reset
      indsUp[j, :] = inds[j, :]

  def saveVTK(self, fname):
    f = open(fname, 'w')
    print >> f, "# vtk Data Version 2.0"
    print >> f, "upwind3.py"
    print >> f, "ASCII"
    print >> f, "DATASET RECTILINEAR_GRID"
    print >> f, "DIMENSIONS", 
    # in VTK the first dimension varies fastest so need 
    # to invert the order of the dimensions
    if self.ndims > 2:
      print >> f, '%d' % (self.numCells[2] + 1),
    else:
      print >> f, " 1",
    if self.ndims > 1:
      print >> f, '%d' % (self.numCells[1] + 1),
    else:
      print >> f, " 1",
    print >> f, '%d' % (self.numCells[0] + 1)
    print >> f, "X_COORDINATES",
    if self.ndims > 2:
      print >> f, "%d double" % (self.numCells[2] + 1)
      for i in range(self.numCells[2] + 1):
        print >> f, " %f" % (0.0 + self.deltas[2] * i),   
    else:
      print >> f, "1 double"
      print >> f, "0.0"
    print >> f, "\nY_COORDINATES",
    if self.ndims > 1:
      print >> f, "%d double" % (self.numCells[1] + 1)
      for i in range(self.numCells[1] + 1): 
        print >> f, " %f" % (0.0 + self.deltas[1] * i),  
    else:
      print >> f, "1 double"
      print >> f, "0.0"
    print >> f, "\nZ_COORDINATES",
    print >> f, "%d double" % (self.numCells[0] + 1)
    for i in range(self.numCells[0] + 1):
      print >> f, " %f" % (0.0 + self.deltas[0] * i),
    print >> f, "\nCELL_DATA %d" % self.ntot
    print >> f, "SCALARS f double 1"
    print >> f, "LOOKUP_TABLE default"
    for i in range(self.ntot):
      print >> f, self.f[i],
      if (i + 1) % 10 == 0: 
        print >> f
    print >> f
    f.close()

  def checksum(self):
    return numpy.sum(self.f)

  def printOut(self):
    for i in range(len(self.f)):
      print i, ' ', self.f[i]

############################################################################################################
def main():
  import sys

  if len(sys.argv) <= 1:
    print "must specify number of cells in each direction."
    return sys.exit(1)

  ndims = 3
  numCells = [int(sys.argv[1])] * 3
  print "number of cells: ", numCells

  numTimeSteps = 100
  if len(sys.argv) > 2:
    numTimeSteps = int(sys.argv[2])


  velocity = numpy.ones( (ndims,), numpy.float64 )
  lengths = numpy.ones( (ndims,), numpy.float64 )

  # compute dt 
  courant = 0.1
  dt = float('inf')
  for j in range(ndims):
    dx = lengths[j]/ float(numCells[j])
    dt = min(courant * dx / velocity[j], dt)

  up = Upwind(velocity, lengths, numCells)
  #up.saveVTK("up0.vtk")
  for i in range(numTimeSteps):
    up.advect(dt)
    #if i % 10 == 0:
    #up.saveVTK("up" + str(i) + ".vtk")

  #up.printOut()
  print "check sum: ", up.checksum()
  up.saveVTK("up.vtk")

if __name__ == '__main__': main()
